src/fontra/views/fontinfo/panel-sources.js

```diff
@@ -1,14 +1,642 @@
+import { addStyleSheet } from "../core/html-utils.js";
+import { ObservableController } from "../core/observable-object.js";
+import {
+  OptionalNumberFormatter,
+  labelForElement,
+  labeledCheckbox,
+  labeledTextInput,
+  textInput,
+} from "../core/ui-utils.js";
+import { round } from "../core/utils.js";
 import { BaseInfoPanel } from "./panel-base.js";
+import { locationToString, makeSparseLocation } from "/core/var-model.js";
+import "/web-components/add-remove-buttons.js";
+import "/web-components/designspace-location.js";
+import { dialogSetup, message } from "/web-components/modal-dialog.js";
 
 export class SourcesPanel extends BaseInfoPanel {
   static title = "sources.title";
   static id = "sources-panel";
   static fontAttributes = ["axes", "sources"];
 
-  setupUI() {
+  async setupUI() {
+    const sources = await this.fontController.getSources();
+    const fontAxes = this.fontController.axes.axes;
+
+    const container = html.div({
+      style: "display: grid; gap: 0.5em;",
+    });
+
+    // TODO: maybe sort sources by axes and location values
+    for (const identifier of Object.keys(sources)) {
+      container.appendChild(
+        new SourceBox(
+          fontAxes,
+          sources,
+          identifier,
+          this.postChange.bind(this),
+          this.setupUI.bind(this)
+        )
+      );
+    }
+
+    this.panelElement.innerHTML = "";
+    this.panelElement.style = `
+    gap: 1em;
+    `;
     this.panelElement.appendChild(
-      html.div({}, [`⚠️ under construction: placeholder for ${this.constructor.id}`])
+      html.input({
+        type: "button",
+        style: `justify-self: start;`,
+        value: "New font source...",
+        onclick: (event) => this.newSource(),
+      })
+    );
+    this.panelElement.appendChild(container);
+    this.panelElement.focus();
+  }
+
+  async newSource() {
+    const newSource = await this._sourcePropertiesRunDialog();
+    if (!newSource) {
+      return;
+    }
+
+    const undoLabel = `add source '${newSource.name}'`;
+    const sourceIdentifier = newSource.name;
+    // TODO: Maybe use proper sourceIdentifier, not source name
+    const root = { sources: this.fontController.sources };
+    const changes = recordChanges(root, (root) => {
+      root.sources[sourceIdentifier] = newSource;
+    });
+    if (changes.hasChange) {
+      this.postChange(changes.change, changes.rollbackChange, undoLabel);
+      this.setupUI();
+    }
+  }
+
+  async _sourcePropertiesRunDialog() {
+    const sources = await this.fontController.getSources();
+    const locationAxes = this.fontController.axes.axes;
+    const validateInput = () => {
+      const warnings = [];
+      const editedSourceName = nameController.model.sourceName;
+      if (!editedSourceName.length || !editedSourceName.trim()) {
+        warnings.push("⚠️ The source name must not be empty");
+      }
+      if (
+        Object.keys(sources)
+          .map(function (sourceIdentifier) {
```

Comment:
Please use an arrow function instead, because that is how we always spell anonymous funcs in Fontra.

Reply:
ok. done.